ghost/mailgun-client/lib/mailgun-client.js

```diff
@@ -132,16 +141,14 @@ module.exports = class MailgunClient {
             debug(`fetchEvents: finished fetching first page with ${events.length} events`);
 
             let eventCount = 0;
+            const beginTimestamp = mailgunOptions.begin ? Math.ceil(mailgunOptions.begin * 1000) : undefined; // ceil here if we have rounding errors
 
-            pagesLoop:
             while (events.length !== 0) {
-                const batchResult = await batchHandler(events);
-
-                result = result.concat(batchResult);
+                await batchHandler(events);
                 eventCount += events.length;
 
-                if (eventCount >= maxEvents) {
-                    break pagesLoop;
+                if (eventCount >= maxEvents && (!beginTimestamp || !events[events.length - 1].timestamp || (events[events.length - 1].timestamp.getTime() > beginTimestamp))) {
```

Comment:
This fixes the deadlock issue, if maxEvents is 600, and we receive 600 events with the same timestamp as the begin timestamp, we would keep fetching them over and over again.

Reply:
Luckily, Mailgun timestamps have ms precision, so this shouldn't happen a lot + if it happens it should be limited in extra events